active/0024-opentelemetry-trace-integration.md

```diff
@@ -0,0 +1,211 @@
+Obviously, this works fine when function calls are being traced within the context of the same process and needs little efforts when the context is to be propagated to a new spawned process.
+
+However, this approach is not absolutely suitable for EMQX distributed architecture:
+
+- correlated spans can be executed on different nodes and/or by different processes
+- a batch of items relating to different traces can be processed together as a single unit of work, e.g., `emqx_connection`, `emqx_channel` modules process deliver messages in batches, where each message would have a unique trace ID if tracing is enabled.
+
+That’s why the proposed implementation mostly relies on propagating the tracing context as a part of the message itself, which has the following advantages:
+
+- inter-cluster communication doesn’t require any changes to support trace context propagation and is backward compatible (trace context is added to a reserved `#emqx_message.extra` field)
+- tracing individual messages processed in batches is possible and doesn’t require any significant changes in the current implementation.
+
+API and context propagation examples (see: [full draft implementation](https://github.com/emqx/emqx/pull/11696/files#diff-a0fac912929fe4c3aa6b99a65a070f1886f75b2586beb4a7e21d8329be9aaa32)):
+
+```erlang
+-spec trace_publish(Msg, fun((Msg) -> Res)) -> Res when
+    Msg :: emqx_types:message(),
+    Res :: emqx_types:publish_result().
+trace_publish(Msg, PublishFun) ->
+    ...
+
+-spec trace_dispatch(Topic, Msg, fun((Topic, Msg) -> Res)) -> Res when
+    Topic :: emqx_types:topic(),
+    Msg :: emqx_types:message(),
+    Res :: emqx_types:deliver_result().
+trace_dispatch(Topic, Msg, DispatchFun) ->
+   ...
+
+put_ctx_to_msg(OtelCtx, Msg = #message{extra = Extra}) ->
+    Msg#message{extra = lists:keystore(?EMQX_OTEL_CTX, 1, Extra, {?EMQX_OTEL_CTX, OtelCtx})}.
+
+get_ctx_from_msg(#message{extra = Extra}) ->
+    proplists:get_value(?EMQX_OTEL_CTX, Extra).
+```
+
+Some drawbacks of the proposed implementation should also be mentioned:
+
+- internal tracing API (as of now, implemented in `emqx_otel_trace` module) is not decoupled from the rest of the code base: each traceable action (span) is traced by a specific function and all these functions are quite specific. For example, they may extract/propagate the context differently and/or rely on the previous (parent) span. For now, it doesn’t seem feasible to create a generic trace wrapper that can trace an arbitrary function.
+- current implementation modified `mqtt_packet` record by adding an `internal_extra` field to trace some events occurring after `emqx_message` is converted to `mqtt_packet`:
+
+  ```erlang
+  -record(mqtt_packet, {
+    ...
+    %% Experimental, added to carry Opentelemetry context after message is converted to packet,
+    %% must always be ignored when serialized, as it is not a part of MQTT packet
+    internal_extra = [] :: list()
```

Comment:
can we keep tracing in user properties ?
i.e. `#mqtt_packet.variable`

Reply:
Yes, but if so, I think we would mix our custom internal data (that shouldn't be serialized) together with MQTT specific properties :thinking:
Another idea I had is to wrap `mqtt_packet` record with some additional structure. Can be as simple as: `{#mqtt_packet{}, Extra}` but it might be error prone.

Reply:
Changed,  `internal_extra` field in `properties` map will be excluded when serialized to frame:
```
serialize_property(_, Disabled) when Disabled =:= disabled; Disabled =:= undefined ->
    <<>>;
serialize_property(internal_extra, _) ->
    <<>>;
serialize_property('Payload-Format-Indicator', Val) ->
    <<16#01, Val>>;
...
```